syntax = "proto3";

package juju.client.simple.v1alpha1;

import "google/api/annotations.proto";

service SimpleService {
    rpc Status(StatusRequest) returns (StatusResponse) {
        option (google.api.http) = {
            get: "/v1alpha1/status"
        };
    }
    rpc Deploy(DeployRequest) returns (DeployResponse) {
        option (google.api.http) = {
            post: "/v1alpha1/deploy"
            body: "*"
        };
    }
    rpc RemoveApplication(RemoveApplicationRequest) returns (RemoveApplicationResponse) {
        option (google.api.http) = {
            post: "/v1alpha1/removeApplication"
            body: "*"
        };
    }
}

message StatusRequest {
    repeated string patterns = 1;
}

message StatusResponse {
    ModelStatusInfo model = 1;
    
}

message ModelStatusInfo {
    string name = 1;
    string type = 2;
    map<string, ApplicationStatus> applications = 3;
}

message ApplicationStatus {
    string charm_name = 1;
    string charm_version = 2;
    DetailedStatus status = 3;
}

message DetailedStatus {
    string status = 1;
    string info = 2;
    string life = 3;
}

message DeployRequest {
    string charm_name = 1;
    string application_name = 2;
    uint32 num_units = 3;
}

message DeployResponse {
}

message RemoveApplicationRequest {
    string application_name = 1;
    bool destroy_storage = 2;
    bool force = 3;
}

message RemoveApplicationResponse {
    repeated string detached_storage_tags = 1;
    repeated string destroyed_storage_tags = 2;
    repeated string destroyed_uni_tagss = 3;
}
